Library/Homebrew/utils/shfmt.sh

```diff
@@ -56,4 +69,155 @@ then
+  # )
+  # for var in "${ARRAY[@]}"
+  # do
+
+  # TODO: use bash built-in regex match syntax instead
+  if grep -qE '^\s*for .*\\\(#.*\)\?$' "${file}"
+  then
+    # TODO: add line number
+    onoe "Multi-line for statement detected."
+    return 1
+  fi
+}
+
+check_IFS_newline() {
+  local file="$1"
+  check_read_and_write "${file}" || return 1
+
+  # check pattern:
+  # IFS=$'\n'
+  #
+  # use the following instead
+  # while read -r line
+  # do
+  # done < <(command)
+
+  # TODO: use bash built-in regex match syntax instead
+  if grep -qE "^[^#]*IFS=\\\$'\\\\n'" "${file}"
+  then
+    # TODO: add line number
+    onoe "Pattern \`IFS=\$'\\\\n'\` detected."
+    return 1
+  fi
+}
+
+wrap_then() {
+  local file="$1"
+  check_read_and_write "${file}" || return 1
+
+  # TODO: wrap `then` to a separated line
+  # before:                   after:
+  # if [[ ... ]]; then        if [[ ... ]]
+  #                           then
+  #
+  # before:                   after:
+  # if [[ ... ]] ||           if [[ ... ]] ||
+  #   [[ ... ]]; then           [[ ... ]]
+  #                           then
+
+  true
+}
+
+# Probably merge into the above function
+wrap_do() {
+  local file="$1"
+  check_read_and_write "${file}" || return 1
+
+  # TODO: wrap `do` to a separated line
+  # before:                   after:
+  # for var in ...; do        for var in ...do
+  #                           do
+
+  true
+}
+
+align_multiline_if_condition() {
+  local file="$1"
+  check_read_and_write "${file}" || return 1
+
+  # TODO: align multiline if condition (indent with 3 spaces or 6 spaces (start with "-"))
+  # before:                   after:
+  # if [[ ... ]] ||           if [[ ... ]] ||
+  #   [[ ... ]]                  [[ ... ]]
+  # then                      then
+  #
+  # before:                   after:
+  # if [[ -n ... || \           if [[ -n ... || \
+  #   -n ... ]]                       -n ... ]]
+  # then                        then
+
+  true
+}
+
+# TODO: it's hard to align multiline switch cases
+align_multiline_switch_cases() {
+  true
+}
+
+format() {
+  local file="$1"
+  local tempfile="${file%/*}/.${file#*/}.temp"
+  local retcode=0
+
+  cp -af "${file}" "${tempfile}"
+
+  # Format with `shfmt` first
+  "${SHFMT}" -w "${SHFMT_ARGS[@]}" "${tempfile}"
+
+  # Fail fast when forbidden patterns detected
+  if check_multiline_for_statements "${tempfile}" ||
+     check_IFS_newline "${tempfile}"
+  then
+    rm -f "${tempfile}" 2>/dev/null
+    return 1
+  fi
+
+  # Tweak it with custom shell script styles
+  wrap_then "${tempfile}"
+  wrap_do "${tempfile}"
+  align_multiline_if_condition "${tempfile}"
+
+  if ! diff -q "${file}" "${tempfile}"
+  then
+    # Show differences
+    diff "${file}" "${tempfile}"
+    if [[ -n "${INPLACE}" ]]; then
+      cp -af "${tempfile}" "${file}"
+    fi
+    retcode=1
+  else
+    # File is identical between code formations (good styling)
+    retcode=0
+  fi
+  rm -f "${tempfile}" 2>/dev/null
+  return "${retcode}"
+}
```

Comment:
I write a prototype here to run `shfmt` first then tweak it with custom shell styles. Should we implement these in `ruby` rather than shell script?